flutter/lib/src/span_frame_metrics_collector.dart

```diff
@@ -0,0 +1,212 @@
+  /// and exists for the duration of the application's lifetime.
+  ///
+  /// Stopping the frame tracking means setting [isTrackingPaused] is `true`
+  /// to prevent actions being done when the frame callback is triggered.
+  void startFrameTracking() {
+    _isTrackingPaused = false;
+
+    if (!_isTrackingRegistered) {
+      _frameCallbackHandler?.addPersistentFrameCallback(measureFrameDuration);
+      _isTrackingRegistered = true;
+    }
+  }
+
+  /// Records the duration of a single frame and stores it in [frameDurations].
+  ///
+  /// This method is called for each frame when frame tracking is active.
+  void measureFrameDuration(Duration duration) async {
+    if (_isTrackingPaused) return;
+
+    if (!_stopwatch.isRunning) {
+      _stopwatch.start();
+    }
+
+    await _frameCallbackHandler?.endOfFrame;
+
+    final frameDuration = _stopwatch.elapsedMilliseconds;
+    // ignore: invalid_use_of_internal_member
+    frameDurations[getUtcDateTime()] = frameDuration;
+
+    _stopwatch.reset();
+
+    if (_frameCallbackHandler?.hasScheduledFrame == true) {
+      _stopwatch.start();
+    }
+  }
+
+  Future<void> recordSpanFrameMetrics(
+      ISentrySpan span, DateTime endTimestamp) async {
+    final displayRefreshRate = await _native?.displayRefreshRate();
+    if (displayRefreshRate == null) {
+      options.logger(SentryLevel.warning,
+          'Display refresh rate is not available. Dropping the frame metrics');
+      clear();
+      return;
+    }
+
+    activeSpans.removeWhere(
+        (element) => element.context.spanId == span.context.spanId);
+
+    final frameMetrics =
+        calculateFrameMetrics(span, endTimestamp, displayRefreshRate);
+
+    frameMetrics.forEach((key, value) {
+      span.setData(key, value);
+    });
+
+    // This will call the methods on the tracer, not on the span directly
+    if (span is SentrySpan && span.isRootSpan) {
+      frameMetrics.forEach((key, value) {
+        // ignore: invalid_use_of_internal_member
+        span.tracer.setData(key, value);
+      });
+      frameMetrics.forEach((key, value) {
+        // In measurements we change e.g frames.total to frames_total
+        span.setMeasurement(key.replaceAll('.', '_'), value);
+      });
+    }
+  }
+
+  Map<String, int> calculateFrameMetrics(
+      ISentrySpan span, DateTime endTimestamp, int displayRefreshRate) {
+    const frozenFrameThresholdMs = 700;
+    final expectedFrameDuration = ((1 / displayRefreshRate) * 1000).toInt();
+
+    // Filter frame durations within the span's time range
+    final timestamps = frameDurations.keys
+        .takeWhile((value) =>
+            value.isBefore(endTimestamp) && value.isAfter(span.startTimestamp))
+        .toList();
+
+    if (timestamps.isEmpty) {
+      options.logger(
+          SentryLevel.info, 'No frame durations available in frame tracker.');
+      return {};
+    }
+
+    final slowFrames = timestamps.where((timestamp) {
```

Comment:
I understand the logic here, but we are iterating on the frames 3 times, plus iterating on the filtered lists.
We can do everything in one for loop, just adding the current frame duration to some local variables